### PR TITLE
feat: multi-key combo remapping (Listen capture)

### PR DESCRIPTION
**Pull request by BunnyTaylor**
## Summary

Lets a single physical key be remapped to a **modifier+key combo** (e.g. `Ctrl+Shift+Alt+P`) instead of just a single key. Previously the picker only allowed selecting one target key.

A new **Listen** button captures whatever combo the user presses on their keyboard and encodes it directly into the existing `FirmwareCode` format the firmware already understands.

## Why this is a small change

The firmware already supports modifier+key combos as a single 24-bit value:
- bits 23..16 = modifier flag bitmask (LCtrl `0x01` … RWin `0x80`)
- bits 15..8  = USB HID keycode of the base key

Existing macro entries like `0x010400` (Ctrl+A) in [`KEY_MAP`](https://github.com/vinc3m1/kludgeknight/blob/main/src/types/keycode.ts) prove the firmware accepts these. `BufferCodec` already writes arbitrary 24-bit values per key slot, so **no protocol, device, or HID-layer changes were needed** — this is purely UI + encoding helpers.

## What changed

**New files**
- `src/components/KeyComboCapture.tsx` — Listen widget. Three-state machine (`idle → listening → captured`); attaches window-level `keydown`/`keyup` capture-phase listeners with `preventDefault()` while listening.
- `src/utils/eventCodeToVk.ts` — maps browser `KeyboardEvent.code` (e.g. `KeyP`, `F4`, `Numpad7`) → Windows VK codes. Modifier codes intentionally excluded; they're tracked as held state.

**Edited**
- `src/types/keycode.ts` — exported `MOD_LCTRL` … `MOD_RWIN`, `MOD_MASK`, `HID_MASK`. Added pure helpers: `combineModifiersAndKey()`, `firmwareCodeToParts()`, `firmwareCodeToLabel()`, `firmwareCodeBaseLabel()`, `firmwareCodeHasModifiers()`.
- `src/types/keycode.test.ts` — 12 new unit tests covering encode/decode, L vs R modifier disambiguation, and exact-match precedence (so `Fn` `0xb000` and existing macro entries don't get re-decomposed).
- `src/components/KeyRemapper.tsx` — added the Listen capture above the existing click-to-pick picker (single-key picker stays as fallback). Bumped fw hex display from 4 to 6 chars so combos render in full.
- `src/components/KeyboardCanvas.tsx` — combo-mapped keys now render with a distinct amber tint, a small `+` SVG marker, and a `<title>` tooltip with the full combo label. Added a legend entry.
- `src/utils/localeLabels.ts` — `getLocalizedKeyName()` decomposes unmapped codes with modifier bits set into a readable label (`Ctrl+Shift+Alt+P`) instead of dumping hex.

## Implementation notes

- Uses `event.code` (layout-independent physical key) rather than `event.key`, since the firmware speaks in physical-key terms.
- L vs R modifier distinction comes from `event.code` (`ControlLeft` vs `ControlRight`), not `event.ctrlKey`.
- `Meta` keys (Cmd on macOS, Win on Windows/Linux) map to the firmware's Win modifier flags.
- A few OS-level shortcuts (`Ctrl+T`, `F11`, `Cmd+Q`) cannot be intercepted by the browser; the UI hint mentions this.

## Test plan

- [x] `bun run validate` passes locally (lint + typecheck + tests + build) — 20 tests pass, 0 warnings.
- [x] 12 new unit tests cover combo encoding/decoding edge cases.
- [x] Manual: connect RK keyboard → click a key → click Listen → press `Ctrl+Shift+Alt+P` → Apply → reload → mapping persists.
- [x] Hardware: physical key remapped to combo emits the correct combo.

## Demo video

https://github.com/user-attachments/assets/07530823-2fbb-47b2-80a2-2366647a3818

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

**Comment by irishman2020** [comment]
Does this need to be tested more before release?

**Comment by BunnyTaylor** [comment]
> Does this need to be tested more before release?

I've been using it frequently on my own device, so thats the only testing ive done, plus whats shown in the video. However maybe testing on more keyboards would help. I only have the RK S70 keyboard.

**Comment by irishman2020** [comment]
That's the only one that matters lol. Hopefully @vinc3m1 will merge this change in to the githubpages version.

Unrelated: Have you done any testing with the lighting controls? I noticed that mine only updates the left side of the keyboard.

File: src/components/KeyboardCanvas.tsx [hunk 1]
```diff
@@ -50,24 +51,53 @@ export function KeyboardCanvas({ device, onKeyClick, selectedKeyIndex, imageMani
   };
 
   const renderKey = (key: Key, index: number) => {
-    const hasMapping = device.hasMapping(key.bIndex);
+    const mapping = device.getMapping(key.bIndex);
```
Comment (vinc3m1) [inline src/components/KeyboardCanvas.tsx:54]: **Should-fix: this re-skins existing modifier/macro remaps as amber combos.**

`firmwareCodeHasModifiers` returns true for several entries already in `KEY_MAP`:

- Single-modifier remaps: `Left Shift = 0x020000`, `Right Ctrl = 0x100000`, `Left Win = 0x080000`, etc. (`src/types/keycode.ts:139-148`)
- Macro entries: `Select All (Ctrl+A) = 0x010400`, `Copy = 0x010600`, `Paste = 0x011900`, etc. (`src/types/keycode.ts:174-181`)

So anyone who had remapped a key to "Left Shift" or "Select All" today (green) will see amber + `+` marker after this PR. The tooltip stays correct via label preservation, but the color/marker silently shifts. For the macros it's arguably accurate; for a lone modifier remap it's a visual regression — a single Shift isn't a combo.

Suggested fix: gate on "has a base key AND not an exact `KEY_MAP` entry" so existing semantics are preserved and only newly-captured combos get the amber styling:

```tsx
const isCombo = hasMapping
  && firmwareCodeHasModifiers(mapping)
  && (mapping & 0x00ff00) !== 0
  && !Object.values(KEY_MAP).some(k => k.fw === mapping);
```

---
_Generated by [Claude Code](https://claude.ai/code)_

File: src/components/KeyComboCapture.tsx [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+      e.stopPropagation();
+
+      if (isModifierEventCode(e.code)) {
+        const flag = modifierFlagForEventCode(e.code);
+        setState(prev =>
+          prev.kind === 'listening' ? { ...prev, held: prev.held | flag } : prev
+        );
+        return;
+      }
+
```
Comment (vinc3m1) [inline src/components/KeyComboCapture.tsx:81]: **Should-fix: capture can leave the badge inconsistent with what the user pressed.**

Two coupled issues here:

1. `onChange(fw)` lives inside the `setState` updater — updaters must be pure (React StrictMode invokes them twice in dev), and this also makes the side effect depend on the updater being called the right number of times.
2. The `keydown` listener stays attached until *this* effect's cleanup runs, which is after React commits. In that sub-ms window a stray `keydown` (auto-repeat, or a second key mashed near-simultaneously) re-enters the handler with `prev.kind === 'captured'`, falls through to `heldNow = 0`, and clobbers the captured combo with just the base key. Tiny window, but the badge ends up showing the wrong key.

Robust fix: attach with `AbortController` so we can detach **synchronously** in the same event tick as capture, and pull the `onChange` out of the updater. Sketch:

```tsx
const heldRef = useRef(0);

useEffect(() => {
  if (state.kind !== 'listening') return;
  const controller = new AbortController();
  const { signal } = controller;

  const onKeyDown = (e: KeyboardEvent) => {
    e.preventDefault();
    e.stopPropagation();

    if (isModifierEventCode(e.code)) {
      heldRef.current |= modifierFlagForEventCode(e.code);
      setState({ kind: 'listening', held: heldRef.current });
      return;
    }
    const vk = EVENT_CODE_TO_VK[e.code];
    if (vk == null) return;
    const baseFw = vkToFirmwareCode(vk);
    if (baseFw == null) return;

    controller.abort(); // detaches both listeners synchronously
    const fw = combineModifiersAndKey(heldRef.current, baseFw);
    setState({ kind: 'captured', fw });
    onChange(fw);
  };

  const onKeyUp = (e: KeyboardEvent) => {
    if (isModifierEventCode(e.code)) {
      heldRef.current &= ~modifierFlagForEventCode(e.code);
      setState({ kind: 'listening', held: heldRef.current });
    }
  };

  window.addEventListener('keydown', onKeyDown, { capture: true, signal });
  window.addEventListener('keyup', onKeyUp, { capture: true, signal });
  return () => controller.abort();
}, [state.kind, onChange]);
```

Also reset `heldRef.current = 0` in `startListening`. `AbortSignal` on `addEventListener` is supported everywhere WebHID runs, so no compat concern.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: src/types/keycode.ts [hunk 3]
```diff
@@ -268,6 +292,106 @@ export function getKeysByCategory(category: KeyCategory): KeyInfo[] {
...
+ */
+export function firmwareCodeBaseLabel(fw: FirmwareCode): string {
+  const exact = findKeyInfoByFw(fw);
+  if (exact && !firmwareCodeHasModifiers(fw)) return exact.label;
+
+  const { baseFw } = firmwareCodeToParts(fw);
+  if (baseFw === 0) {
+    // No base key — fall back to full label (covers pure-modifier remaps).
+    return firmwareCodeToLabel(fw);
+  }
```
Comment (vinc3m1) [inline src/types/keycode.ts:390]: Nit: `firmwareCodeBaseLabel` is exported and tested but not consumed by any caller in the diff. Either wire it up or drop it — dead exports rot.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: src/components/KeyRemapper.tsx [hunk 4]
```diff
@@ -455,9 +456,18 @@ export function KeyRemapper({ device, imageManifest }: KeyRemapperProps) {
...
+              <KeyComboCapture
+                initialValue={selectedTargetKey ?? undefined}
+                onChange={setSelectedTargetKey}
+              />
+            </div>
+
             {/* Main Keyboard Layout - Only this scrolls horizontally */}
             <div>
-              <h4 className="text-sm font-semibold mb-3 text-card-foreground">Standard Layout</h4>
+              <h4 className="text-sm font-semibold mb-3 text-card-foreground">Or pick a single key</h4>
```
Comment (vinc3m1) [inline src/components/KeyRemapper.tsx:470]: Nit: heading is slightly misleading — the picker also lets you select modifiers and macro keys (e.g. "Select All (Ctrl+A)") which are themselves combos at the firmware level. Something like "Or pick from the list" would be more accurate.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: src/components/KeyComboCapture.tsx [hunk 5]
```diff
@@ -0,0 +1,154 @@
...
+  modifierFlagForEventCode,
+} from '../utils/eventCodeToVk';
+
+interface KeyComboCaptureProps {
+  initialValue?: FirmwareCode;
+  onChange: (fw: FirmwareCode) => void;
+}
+
+type CaptureState =
+  | { kind: 'idle'; held: number }
```
Comment (vinc3m1) [inline src/components/KeyComboCapture.tsx:22]: Nit: `idle`'s `held` is always `0` — can be simplified to `{ kind: 'idle' }`.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Review by vinc3m1 — COMMENTED** [review]
Nice scoping — the encoding observation (firmware already accepts 24-bit modifier+key) is the right insight and keeps the change off the HID/BufferCodec/device layers. Tests on the encode/decode helpers are solid.

Two should-fix items inline:

1. **Canvas tinting regression** — `firmwareCodeHasModifiers` matches existing `KEY_MAP` entries (single-modifier remaps like `Left Shift = 0x020000`, and macro entries like `Select All = 0x010400`), so users who previously remapped to those will silently shift from green to amber + "+" marker. Suggested guard inline.
2. **KeyComboCapture cleanup race** — listener cleanup runs after React commits, so a stray `keydown` in that sub-ms window can clobber the captured combo with just the base key, leaving the badge inconsistent with what the user pressed. `AbortController` makes the detach synchronous; also moves the `onChange` side effect out of the `setState` updater. Sketch inline.

Plus three nits (dead helper, misleading heading, redundant state field).

Not blocking, just observations:
- PR is well-isolated — no protocol/HID/storage changes.
- The combo capture section sits above the single-key picker; fine if combos are intended to be the primary flow, but worth considering whether to put it below or behind a disclosure to keep the default UX lean.
- Cross-flow state sync (picker updates Listen badge and vice versa) is intentional and reads well.

---
_Generated by [Claude Code](https://claude.ai/code)_
